contributors/design-proposals/deployment.md

```diff
@@ -116,35 +118,50 @@ type DeploymentStatus struct {
-   new RC).
+   new RS).
    - The label key will be "pod-template-hash".
-   - The label value will be hash of the podTemplateSpec for that RC without
-     this label. This value will be unique for all RCs, since PodTemplateSpec should be unique.
-   - If the RCs and pods dont already have this label and selector:
-     - We will first add this to RC.PodTemplateSpec.Metadata.Labels for all RCs to
+   - The label value will be hash of the podTemplateSpec for that RS without
+     this label. This value will be unique for all RSs, since PodTemplateSpec should be unique.
+   - If the RSs and pods dont already have this label and selector:
+     - We will first add this to RS.PodTemplateSpec.Metadata.Labels for all RSs to
        ensure that all new pods that they create will have this label.
-     - Then we will add this label to their existing pods and then add this as a selector
-       to that RC.
-3. Find if there exists an RC for which value of "pod-template-hash" label
-   is same as hash of DeploymentSpec.PodTemplateSpec. If it exists already, then
-   this is the RC that will be ramped up. If there is no such RC, then we create
-   a new one using DeploymentSpec and then add a "pod-template-hash" label
-   to it. RCSpec.replicas = 0 for a newly created RC.
-4. Scale up the new RC and scale down the olds ones as per the DeploymentStrategy.
-   - Raise an event if we detect an error, like new pods failing to come up.
-5. Go back to step 1 unless the new RC has been ramped up to desired replicas
-   and the old RCs have been ramped down to 0.
-6. Cleanup.
+     - Then we will add this label to their existing pods
+     - Eventually we flip the RS selector to use the new label.
+     This process potentially can be abstracted to a new endpoint for controllers.
+   - Hashing an API object such as PodTemplateSpec means that the resulting hash is subject to
+     change due to API changes in PodTemplateSpec between minor versions of Kubernetes when new
+     API changes are introduced. A new API field will be introduced in DeploymentSpec called
+     TemplateGeneration. The new field will be a monotonic counter that will track the number of
+     times an update happened in the PodTemplateSpec of the Deployment, similarly to how
+     metadata.Generation tracks updates in the Spec of all first-class API objects.
+     The naming scheme used by ReplicaSets (*deployment.name-podtemplatehash*) will need to change
+     to something different because we cannot migrate old ReplicaSets to use TemplateGeneration
+     for their names and we want to avoid name collisions. For new RS names, we can either
+     append the TemplateGeneration to the Deployment name or we can compute the hash of
+     *deployment.name+deployment.templategeneration* and extend its size in different ways to
+     ensure that a new RS name will never collide with an old one.
+     Since owner references have been added in Pods since 1.4, we should be able to stop adding
```

Comment:
Thinking about this more, now I think not using the extra label is dangerous because the Pods owned by a ReplicaSet will be claimable by all existing ReplicaSets under a Deployment in the event of a non-cascading deletion of the owner ReplicaSet.